src/NUnitFramework/framework/Constraints/UniqueItemsConstraint.cs

```diff
@@ -60,5 +69,123 @@ protected override bool Matches(IEnumerable actual)
 
             return true;
         }
+
+        private bool? TryFastAlgorithm(IEnumerable actual)
+        {
+            // If the user specified any external comparer with Using, exit
+            if (UsingExternalComparer)
+                return null;
+
+            // If IEnumerable<T> is not implemented exit,
+            // Otherwise return value is the Type of T
+            Type memberType = GetGenericTypeArgument(actual);
+            if (memberType == null || IsHandledSpeciallyByNUnit(memberType))
+                return null;
+
+            string methodName = nameof(ItemsUnique);
+            bool makeGeneric = true;
+
+            // Special handling for ignore case with strings and chars
+            if (IgnoringCase)
+            {
+                if (memberType == typeof(string))
+                {
+                    methodName = nameof(StringsUniqueIgnoringCase);
+                    makeGeneric = false;
+                }
+                else
+                if (memberType == typeof(char))
+                {
+                    methodName = nameof(CharsUniqueIgnoringCase);
+                    makeGeneric = false;
+                }
+            }
+
+            MethodInfo method = GetType().GetMethod(methodName, BindingFlags.Static | BindingFlags.NonPublic);
+            if (method == null)
+                throw new InvalidOperationException("Internal error - Method not found: " + methodName);
+            
+            if (makeGeneric)
+                method = method.MakeGenericMethod(memberType);
+
+            return (bool)method.Invoke(null, new object[] { actual });
+        }
+
+        private static bool ItemsUnique<T>(IEnumerable<T> actual)
+        {
+            var hash = new HashSet<T>();
+
+            foreach (T item in actual)
+            {
+                if (hash.Contains(item))
+                    return false;
+
+                hash.Add(item);
+            }
```

Comment:
You can get rid of one of the hash lookups by replacing the loop body with `if (!hash.Add(item)) return false`.

Reply:
Didn't know that!